nixos/modules/services/games/quake3-server.nix

```diff
@@ -0,0 +1,95 @@
+{ config, pkgs, lib, ... }:
+with lib;
+
+let
+  cfg = config.services.quake3-server;
+  configFile = pkgs.writeText "q3ds-extra.cfg" ''
+    set net_port ${builtins.toString cfg.port}
+
+    ${cfg.extraConfig}
+  '';
+  defaultBaseq3 = pkgs.requireFile rec {
+    name = "baseq3";
+    hashMode = "recursive";
+    sha256 = "5dd8ee09eabd45e80450f31d7a8b69b846f59738726929298d8a813ce5725ed3";
+    message = ''
+      Unfortunately, we cannot download ${name} automatically.
+      Please purchase a legitimate copy of Quake 3 and change into the installation directory.
+
+      mkdir /tmp/baseq3
+      cp baseq3/pak*.pk3 /tmp/baseq3
+      nix-store --add-fixed sha256 --recursive /tmp/baseq3
+    '';
+  };
+  home = pkgs.runCommand "quake3-home" {} ''
+      mkdir -p $out/.q3a/baseq3
+
+      for file in ${cfg.baseq3}/*; do
+        ln -s $file $out/.q3a/baseq3/$(basename $file)
+      done
+
+      ln -s ${configFile} $out/.q3a/baseq3/nix.cfg
+  '';
+in {
+  options = {
+    services.quake3-server = {
+      enable = mkEnableOption "Enable the Quake 3 dedicated server";
+
+      port = mkOption {
+        type = types.int;
+        default = 27960;
+        description = ''
+          UDP Port the server should listen on.
+        '';
+      };
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Open the firewall.
+        '';
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Extra configuration options.
+        '';
+      };
+
+      baseq3 = mkOption {
```

Comment:
I'm thinking aloud here... maybe a good idea, maybe a bad idea...

How large is this file/are these files? If they are large this means the user has to duplicate them? Maybe we should change the `type` to `either package path` and allow the sysadmin to directly pass files into the program, or a store path.

Is that a good idea?

Reply:
The baseq3 directory is around 500Mb, duplicating those is a absolute no-go for me. IMO using a path is not as good of an option because - as discussed above - using a path would force us to move the configuration files out of the nix store. Keeping everything on the nix store guarantees that all files are in a state we expect. (There are possibilities to alter config files from the server, currently they just silently fail. If they succeed we have to reset the server configuration each restart to have a reproducible system)

Reply:
What I mean is that you use `requireFile` to get the 500Mb file into the nix store. This means there are now 2 copies on the users computer. A copy in `/tmp` (or wherever they put it) and a copy under `/nix/store`.

... right?

Reply:
Well, the copy in `/tmp` can be deleted after it has been added to the nix store. Also I guess that most deployments of this will be headless servers, where you can `nix-store --add-fixed ...` the files on your desktop and then `nix-copy-closure` the created nix store path to the server.

Reply:
OK. Fair enough.